### PR TITLE
Updates to Sensfunc masking and outputs

### PR DESCRIPTION
**Pull request by tbowers7**
This PR contains two main parts related to sensitivity functions:
1. Updating the masking of standard star spectra prior to bspline fitting.
2. Adding configuration keys and other metadata to the output FITS file headers.
--------

Masking:
* Clean up the list of recombination lines masked (wavelengths now in vacuum), and remove duplicate code from `telluric.py` (all masking is now in `flux_calib.py`).
* Separate out hydrogen and ionized helium recombination lines as separate SensFuncPar parameters, since the latter only matter for very hot standard stars.
* Apply the full `zeropoint_fitmask` when computing the zeropoint (bugfix: was previously only applying `zeropoint_clean_gpm`).

Additional metadata in output files:
* Add information about telescope, instrument, observation datetime, and configuration keys to the sensitivity function output FITS file.
* In `Spectrograph. subheader_for_spec()`, add configuration keys and observation datetime to the FITS header for use downstream.

Other changes:
* Corrected docstrings and added comments as breadcrumbs for the developer (_e.g._, changes to `pypeit/core/fitting.py` are purely commentary and documentary).
* Tweaks to QA / debugging plots for clarity.

### CODE REVIEW

**Comment by profxj** [comment]
Can this PR include/resolve the issue raised on the Users slack
as regards the fluxing docs?

See the post by Peng on March 8

File: pypeit/core/flux_calib.py [hunk 1]
```diff
@@ -638,10 +639,12 @@ def sensfunc(wave, counts, counts_ivar, counts_mask, exptime, airmass, std_dict,
...
             with the shape (norders,) giving the order numbers
-        mask_abs_lines (bool):
-            If True, mask stellar absorption lines before fitting sensitivity function. Default = True
+        mask_hydrogen_lines (bool):
+            If True, mask stellar hydrogen absorption lines before fitting sensitivity function. Default = True
+        mask_helium_lines (bool):
+            If True, mask stellar helium absorption lines before fitting sensitivity function. Default = False
         balm_mask_wid (float):
-            Parameter describing the width of the mask for or stellar absorption lines (i.e. mask_abs_lines=True). A region
+            Parameter describing the width of the mask for or stellar absorption lines (i.e. mask_hydrogen_lines=True). A region
```
Comment (jhennawi) [inline pypeit/core/flux_calib.py:645]: Does this also apply to the width of the helium line mask.

**Review by badpandabear — APPROVED** [review]

File: pypeit/core/flux_calib.py [hunk 2]
```diff
@@ -948,17 +958,17 @@ def fit_zeropoint(wave, Nlam_star, Nlam_ivar_star, gpm_star, std_dict,
...
         plt.legend(fancybox=True, shadow=True)
         plt.show()
 
     return zeropoint_data, zeropoint_data_gpm, zeropoint_fit, zeropoint_fit_gpm
 
 
 def get_mask(wave_star, flux_star, ivar_star, mask_star, 
-             mask_abs_lines=True, mask_telluric=True, balm_mask_wid=10., 
-             trans_thresh=0.9):
+             mask_hydrogen_lines=True, mask_helium_lines=False,
```
Comment (jhennawi) [inline pypeit/core/flux_calib.py:968]: Should we have a helium mask width parameter or is it just using the Balmer_mask_width. If so, maybe document that.

File: pypeit/core/flux_calib.py [hunk 3]
```diff
@@ -1089,7 +1085,78 @@ def get_mask(wave_star, flux_star, ivar_star, mask_star,
         else:
             msgs.info('Your spectrum is bluer than 9100A, only optical telluric regions are masked.')
 
-    return mask_bad, mask_balm, mask_tell
+    return mask_bad, mask_recomb, mask_tell
+
```
Comment (jhennawi) [inline pypeit/core/flux_calib.py:1081]: Thanks for cleaning this up.

File: pypeit/core/flux_calib.py [hunk 4]
```diff
@@ -1012,43 +1026,25 @@ def get_mask(wave_star, flux_star, ivar_star, mask_star,
...
-        lines_pfund = np.array([22788.0, 32961.0, 37395.0, 46525.0,74578.0])
-        for line_pfund in lines_pfund:
-            ipfund = np.abs(wave_star - line_pfund) <= balm_mask_wid
-            mask_balm[ipfund] = False
+    if mask_hydrogen_lines:
+        mask_recomb = mask_stellar_hydrogen(
+            wave_star, mask_width=balm_mask_wid, mask_star=mask_recomb
+        )
+
+    if mask_helium_lines:
```
Comment (jhennawi) [inline pypeit/core/flux_calib.py:1036]: Should we create a mask helium lines function similar to what you did for hydrogen?

File: pypeit/core/telluric.py [hunk 5]
```diff
@@ -1179,63 +1179,8 @@ def eval_poly_model(theta, obj_dict):
...
-    lines_pfund = np.array([22788.0, 32961.0, 37395.0, 46525.0, 74578.0])
-    for line_pfund in lines_pfund:
-        ipfund = np.abs(wave_star - line_pfund) <= mask_width
-        mask_star[ipfund] = False
-
-    return mask_star
-
 def sensfunc_telluric(wave, counts, counts_ivar, counts_mask, exptime, airmass, std_dict,
-                      telgridfile, ech_orders=None, polyorder=8, mask_abs_lines=True,
+                      telgridfile, ech_orders=None, polyorder=8, mask_hydrogen_lines=True,
```
Comment (jhennawi) [inline pypeit/core/telluric.py:1184]: Should be masking helium lines here too?

**Review by jhennawi — APPROVED** [review]
Thanks @tbowers7. I have pretty minor comments. The PR is approved, but please take a look at my comments before merging.

**Comment by profxj** [comment]
Will we want to run the DevSuite here, @tbowers7 ?

**Comment by tbowers7** [comment]
> Will we want to run the DevSuite here, @tbowers7 ?

Yes, after I make the changes @jhennawi requested.  I'll request such on Slack when ready.

File: pypeit/core/flux_calib.py (hunk 1 above)
Comment (tbowers7) [reply pypeit/core/flux_calib.py:645]: Based on a few standard star spectra (and some background reading), I chose to empirically set the helium mask width to 0.5 the hydrogen width without introducing a separate user parameter.  If you think this should be fully adjustable, I can make it so.

File: pypeit/core/flux_calib.py (hunk 2 above)
Comment (tbowers7) [reply pypeit/core/flux_calib.py:968]: Added documentation about the `HeII` mask width being set to 0.5 x the hydrogen width.

File: pypeit/core/flux_calib.py (hunk 4 above)
Comment (tbowers7) [reply pypeit/core/flux_calib.py:1036]: Done.

File: pypeit/core/telluric.py (hunk 5 above)
Comment (tbowers7) [reply pypeit/core/telluric.py:1184]: Yes.  Now using the `flux_calib.get_mask()` function, but excluding any telluric masking.

**Comment by tbowers7** [comment]
> Can this PR include/resolve the issue raised on the Users slack as regards the fluxing docs?
> 
> See the post by Peng on March 8

Fixed on `keck_esi` branch, commit 647064a1741d3c78d452b7fcbf8a0deb262be89a

**Comment by tbowers7** [comment]
> > Can this PR include/resolve the issue raised on the Users slack as regards the fluxing docs?
> > See the post by Peng on March 8
> 
> Fixed on `keck_esi` branch, commit [647064a](https://github.com/pypeit/PypeIt/commit/647064a1741d3c78d452b7fcbf8a0deb262be89a)

Cherry-picked onto this branch.

**Comment by tbowers7** [comment]
Still have some work to do...
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  208 passed, 2 skipped, 35 warnings in 198.33s (0:03:18) ---
--- PYTEST UNIT TESTS PASSED  118 passed, 181 warnings in 901.95s (0:15:01) ---
--- PYTEST VET TESTS FAILED  2 failed, 34 passed, 5 skipped, 5 warnings in 628.25s (0:10:28) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 7/158 TESTS  ---
Failed tests:
    gemini_gmos/GS_HAM_R400_860 pypeit_sensfunc
    keck_deimos/900ZD_LVM_5500 pypeit_sensfunc
    keck_mosfire/Y_long pypeit_sensfunc
    gemini_gnirs/32_SB_SXD pypeit_sensfunc
    gemini_gmos/GS_HAM_R400_700 pypeit_sensfunc
    shane_kast_blue/600_4310_d55 pypeit_sensfunc
    keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
```

**Comment by tbowers7** [comment]
Latest DevSuite run:
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  210 passed, 35 warnings in 203.89s (0:03:23) ---
--- PYTEST UNIT TESTS PASSED  118 passed, 181 warnings in 893.45s (0:14:53) ---
--- PYTEST VET TESTS PASSED  41 passed, 4 warnings in 1209.88s (0:20:09) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 3/179 TESTS  ---
Failed tests:
    keck_mosfire/Y_long pypeit_ql
    ldt_deveny/DV8 pypeit
    keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
Skipped tests:
Testing Started at 2023-04-11T02:10:17.855052
Testing Completed at 2023-04-11T11:07:36.428140
Total Time: 8:57:18.573088
```

I think @profxj said the QL test failures will be fixed in an upcoming PR.  The DeVeny test passed on my local machine... the error in the #cloud was related to flexure:
```
Error Messages:

End of Log:
  File "/PypeIt/pypeit/core/flexure.py", line 565, in spec_flex_shift_local
    idx_med_shift = np.where(flex_dict['shift'] == np.median(flex_dict['shift']))[0][0]
IndexError: index 0 is out of bounds for axis 0 with size 0
```

Guidance on next steps appreciated.

**Comment by profxj** [comment]
Ignore the QL tests

The flexure issue, though is odd.  
That line of code seems a bit restrictive.

Thoughts @debora-pe ?

**Comment by debora-pe** [comment]
> Latest DevSuite run:
> 
> ```
> Test Summary
> --------------------------------------------------------
> --- PYTEST PYPEIT UNIT TESTS PASSED  210 passed, 35 warnings in 203.89s (0:03:23) ---
> --- PYTEST UNIT TESTS PASSED  118 passed, 181 warnings in 893.45s (0:14:53) ---
> --- PYTEST VET TESTS PASSED  41 passed, 4 warnings in 1209.88s (0:20:09) ---
> --- PYPEIT DEVELOPMENT SUITE FAILED 3/179 TESTS  ---
> Failed tests:
>     keck_mosfire/Y_long pypeit_ql
>     ldt_deveny/DV8 pypeit
>     keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
> Skipped tests:
> Testing Started at 2023-04-11T02:10:17.855052
> Testing Completed at 2023-04-11T11:07:36.428140
> Total Time: 8:57:18.573088
> ```
> 
> I think @profxj said the QL test failures will be fixed in an upcoming PR. The DeVeny test passed on my local machine... the error in the #cloud was related to flexure:
> 
> ```
> Error Messages:
> 
> End of Log:
>   File "/PypeIt/pypeit/core/flexure.py", line 565, in spec_flex_shift_local
>     idx_med_shift = np.where(flex_dict['shift'] == np.median(flex_dict['shift']))[0][0]
> IndexError: index 0 is out of bounds for axis 0 with size 0
> ```
> 
> Guidance on next steps appreciated.

@tbowers7 @profxj Yes, I think that is a bug that appears only in certain cases. I'll jump on this branch and trying to reproduce the error and fix it. Do you have also a Dev-Suite branch?

**Comment by tbowers7** [comment]
> @tbowers7 @profxj Yes, I think that is a bug that appears only in certain cases. I'll jump on this branch and trying to reproduce the error and fix it. Do you have also a Dev-Suite branch?

@debora-pe I have been running against the `develop` Dev-Suite branch.

**Comment by debora-pe** [comment]
@tbowers7 I have fix the bug. I'm re-running the tests just to make sure everything is still OK.

**Comment by debora-pe** [comment]
@tbowers7 I am getting this error for every keck_deimos dataset in the Dev Suite.  It's not related to this PR, but to your previous PR. Am I missing something?

```
keck_deimos/830G_M_9000_dither pypeit Result: --- FAILED

Logfile:    /Users/dpelliccia/PypeIt/PypeIt-development-suite/REDUX_OUT/keck_deimos/830G_M_9000_dither/keck_deimos_830g_m_9000_dither.test.log
Process Id: 51376
Start time: Tue Apr 11 11:28:57 2023
End time:   Tue Apr 11 11:29:04 2023
Duration:   0:00:06.402008
Mem Usage:  0
Command:    run_pypeit /Users/dpelliccia/PypeIt/PypeIt-development-suite/REDUX_OUT/keck_deimos/830G_M_9000_dither/keck_deimos_830g_m_9000_dither.pypeit -o

Error Messages:

End of Log:
  File "/Users/dpelliccia/PypeIt/PypeIt/pypeit/spectrographs/keck_deimos.py", line 216, in get_detector_par
    measure_files = np.fromiter(
ValueError: cannot create object arrays from iterator
```

**Comment by tbowers7** [comment]
> @tbowers7 I am getting this error for every keck_deimos dataset in the Dev Suite. It's not related to this PR, but to your previous PR. Am I missing something?

@debora-pe Ah, this was a new feature in Numpy 1.23 to [create object arrays from an iterator](https://numpy.org/doc/stable/reference/generated/numpy.fromiter.html) -- if you are using an older version you will get the error.  I will ask on Slack about the dependency version policy.

**Comment by tbowers7** [comment]
Okay, @debora-pe , give it a try now, should work with your version of `numpy`.

**Comment by debora-pe** [comment]
@tbowers7 I ended up upgrading numpy. Thank you for the explanation and for changing the line of code. The tests look good (same as before) except for a `vet` test that does not seem to be related to this PR (maybe I am still missing something).
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  210 passed, 51 warnings in 169.39s (0:02:49) ---
--- PYTEST UNIT TESTS PASSED  118 passed, 191 warnings in 1499.25s (0:24:59) ---
--- PYTEST VET TESTS FAILED  1 failed, 40 passed, 5 warnings in 770.20s (0:12:50) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 2/179 TESTS  ---
Failed tests:
    keck_mosfire/Y_long pypeit_ql
    keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
Skipped tests:
Testing Started at 2023-04-11T13:56:52.032942
Testing Completed at 2023-04-11T22:15:35.887072
Total Time: 8:18:43.854130 
```

```
=========================== short test summary info ============================
FAILED ../../../Pypeit/PypeIt-development-suite/vet_tests/test_datacube.py::test_coadd_datacube
```

**Comment by tbowers7** [comment]
Thanks @debora-pe for your flexure fix.  I got the same test results.

@rcooke-ast , there is a new VET test failure in `test_coadd_datacube`.  The log indicates the issue is that `flux_calib.fit_zeropoint()` is working with an inverse variance array that is completely masked:
```
invvar = masked_array(data=[--, --, --, ..., --, --, --],
```
Is this related to your recent work on `datacube.py`?

**Comment by rcooke-ast** [comment]
> @rcooke-ast , there is a new VET test failure in `test_coadd_datacube`. The log indicates the issue is that `flux_calib.fit_zeropoint()` is working with an inverse variance array that is completely masked:
> 
> ```
> invvar = masked_array(data=[--, --, --, ..., --, --, --],
> ```
> 
> Is this related to your recent work on `datacube.py`?

I don't think so... Just to doubel check, have you merged develop in? There was an error with this recently, but the latest merge where I touched this code, the tests were passing: [PR 1572](https://github.com/pypeit/PypeIt/pull/1572#issuecomment-1500979069)

I'm happy to help solve the current error, though!

**Comment by debora-pe** [comment]
> I don't think so... Just to doubel check, have you merged develop in? There was an error with this recently, but the latest merge where I touched this code, the tests were passing: [PR 1572](https://github.com/pypeit/PypeIt/pull/1572#issuecomment-1500979069)
> 
> I'm happy to help solve the current error, though!

Hi @rcooke-ast . I see in your PR #1572 that you mentioned that you fixed a bug in the vet tests, but I did not find those changes (i.e., I did not find a recent PR in the Dev Suite with these changes). Did you actually merge these changes?

**Comment by rcooke-ast** [comment]
> Hi @rcooke-ast . I see in your PR #1572 that you mentioned that you fixed a bug in the vet tests, but I did not find those changes (i.e., I did not find a recent PR in the Dev Suite with these changes). Did you actually merge these changes?

Sorry for being unclear! The error was in PypeIt (`core.datacube.py`) and not in the vet test in the dev suite. So, I only needed to apply a fix to the pypeit code. As far as I can tell from the code (on the `develop` branch in GitHub), these fixes are incorporated into the `develop` branch of PypeIt.

EDIT :: Also, as far as I can tell, these changes are already merged into this branch. So, I suspect that's not the error here.

**Pull request merged**